test/parallel/test-fs-write-optional-params.js

```diff
@@ -0,0 +1,96 @@
+    { code: expectedCode }
+  );
+  callback(0);
+}
+
+// ([ args ], index, _) signature is for .forEach() or .map()
+function testValidCb([ buffer, options ], index, _, callback) {
+  const dest = path.resolve(tmpdir.path, `rwopt_valid_${index}`);
+  fs.open(dest, 'w+', common.mustSucceed((fd) => {
+    fs.write(fd, buffer, options, common.mustSucceed((bytesWritten, bufferWritten) => {
+      const writeBufCopy = Uint8Array.prototype.slice.call(bufferWritten);
+
+      fs.read(fd, buffer, options, common.mustSucceed((bytesRead, bufferRead) => {
+        const readBufCopy = Uint8Array.prototype.slice.call(bufferRead);
+
+        assert.ok(bytesWritten >= bytesRead);
+        if (options.length !== undefined && options.length !== null) {
+          assert.strictEqual(bytesWritten, options.length);
+        }
+        if (options.offset === undefined || options.offset === 0) {
+          assert.deepStrictEqual(writeBufCopy, readBufCopy);
+        }
+        assert.deepStrictEqual(bufferWritten, bufferRead);
+        fs.close(fd, common.mustSucceed(callback));
+      }));
+    }));
+  }));
+}
+
+const testInvalid = util.promisify(testInvalidCb);
+const testValid = util.promisify(testValidCb);
+
+fs.open(destInvalid, 'w+', common.mustSucceed(async (fd) => {
+  // Test if first argument is not wrongly interpreted as ArrayBufferView|string
+  for (const badBuffer of [
+    undefined, null, true, 42, 42n, Symbol('42'), NaN, [], () => {},
+    Promise.resolve(new Uint8Array(1)),
+    {},
+    { buffer: 'amNotParam' },
+    { string: 'amNotParam' },
+    { buffer: new Uint8Array(1).buffer },
+    new Date(),
+    new String('notPrimitive'),
+    { [Symbol.toPrimitive]: (hint) => 'amObject' },
+
+    // TODO(LiviaMedeiros): add the following after DEP0162 EOL
+    // { toString() { return 'amObject'; } },
+  ]) {
+    await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', badBuffer, {});
+  }
+
+  // First argument (buffer or string) is mandatory
+  await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', undefined, undefined);
+
+  // Various invalid options
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { length: 5 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { offset: 5 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { length: 1, offset: 3 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { length: -1 });
+  await testInvalid(fd, 'ERR_OUT_OF_RANGE', buffer, { offset: -1 });
+  await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', buffer, { offset: false });
+  await testInvalid(fd, 'ERR_INVALID_ARG_TYPE', buffer, { offset: true });
+
+  // Test compatibility with fs.read counterpart
+  await Promise.all([
+    [ buffer, {} ],
+    [ buffer, { length: 1 } ],
+    [ buffer, { position: 5 } ],
+    [ buffer, { length: 1, position: 5 } ],
+    [ buffer, { length: 1, position: -1, offset: 2 } ],
+    [ buffer, { length: null } ],
+    [ buffer, { position: null } ],
+    [ buffer, { offset: 1 } ],
+  ].map(testValid));
```

Comment:
We don't want those to run in parallel, so `Promise.all` is not a great fit.
```suggestion
  for(const [args, index] of [
    [ buffer, {} ],
    [ buffer, { length: 1 } ],
    [ buffer, { position: 5 } ],
    [ buffer, { length: 1, position: 5 } ],
    [ buffer, { length: 1, position: -1, offset: 2 } ],
    [ buffer, { length: null } ],
    [ buffer, { position: null } ],
    [ buffer, { offset: 1 } ],
  ].entries()) {
    await testValid(args, index, undefined);
  }
```
That would also let us simplify the signature of `testValid`.